### DISCUSSION TITLE
URL format differs across components

### DISCUSSION THREAD

**Discussion by max-ostapenko** [quarto-dev/quarto-cli#11398] [Feature Requests]
### Bug description

By default Quarto website uses trailing slash format in URLs and links. But there are many other components that don't always follow this format, e.g.
It makes Google Search crawler remind me about this SEO issue every other week.

### Steps to reproduce

1. Create two files:

- /blog/post1/index.qml
- /blog/post2/index.qml

and **_quarto.yml**:
```
project:
  type: website
  output-dir: _site

website:
  page-navigation: true

  sidebar:
    - contents:
        - auto: blog/**

format:
  html:
    canonical-url: true
```

2. Run `quarto render`
3. Check the links in header elements in `_site/blog/**.html` and `_site/sitemap.xml`.

### Expected behavior

I would expect it to be possible to have unified URLs format experience throughout all the components: URLs, sitemap, and links in the documents, etc.

Those `index.html` need to be cleaned up.

### Actual behavior

In some components places on the website the URL format differs that bring in redirects and results in SEO issues:

1. With `page-navigation: true` the `prev/next` links in the header use URLs with index.html format:
![Screenshot 2024-11-11 at 14 44 02](https://github.com/user-attachments/assets/3230c434-a1bc-4be3-840b-f2864af82b95)

2. `sitemap.xml` also uses `../index.html` format. 

### Quarto check output

```
Quarto 1.5.57
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.57
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Tex:  (not detected)

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.12.5
      Path: /Users/maxostapenko/Documents/GitHub/website-source/.venv/bin/python3
      Jupyter: 5.7.2
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
```

**Comment by cscheid** [comment]
This is complicated. I don't know why Google tooling complains about this, but we can't assume that `foo/` and `foo/index.html` are identical: this behavior is dependent on the web server configuration.

**Comment by max-ostapenko** [comment]
That's exactly the issue - Quarto actually sets trailing slash as canonical URLs (`canonical-url: true`).
But then during the rendering it creates link paths that are not aligned with these.

SEO mentions these because my sitemap.xml says to go to 'https://maxostapenko.com/index.html', but on that page it sees the canonical tag which says 'https://maxostapenko.com/'.
And canonical pages are not listed in sitemap.xml.

Expected behaviour: Quarto aligns its canonical URL path and link format in the rendered HTML and sitemap.

P.S.
IMHO web server configs are not involved here.
I'm currently cleaning the URLs in the sitemap and HTML with the script to fix it.

**Comment by jolars** [comment]
This still doesn't seem to be correctly fixed in v1.9.37. I fixed this with the following script:

```sh
#!/usr/bin/env bash
set -euo pipefail

# Quarto emits sitemap <loc> entries ending in /index.html, while pages'
# rel="canonical" tags point at the trailing-slash form. Google Search Console
# then flags the sitemap URLs as duplicates with no chosen canonical.
# Upstream bug: https://github.com/quarto-dev/quarto-cli/discussions/11398

sitemap="${QUARTO_PROJECT_OUTPUT_DIR:-_site}/sitemap.xml"

[ -f "$sitemap" ] || exit 0

sed -i 's|/index\.html</loc>|/</loc>|g' "$sitemap"
```

Add it to `scripts/fix-sitemap.sh` and set:

```yaml
project:
  post-render:
    - scripts/fix-sitemap.sh
```
